crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,567 @@ impl State {
+    /// assuming that all of the outstanding withdrawal shares will be
+    /// redeemed. The share proceeds are calculated such that the LP share
+    /// price is conserved. When we need to round, we round down to err on
+    /// the side of slightly too few shares being paid out.
+    fn calculate_distribute_excess_idle_share_proceeds(
+        &self,
+        current_block_timestamp: U256,
+        active_lp_total_supply: FixedPoint, // just the number of lp tokens
+        withdrawal_shares_total_supply: FixedPoint,
+        max_share_reserves_delta: FixedPoint,
+        max_iterations: u64,
+    ) -> Result<FixedPoint> {
+        let net_curve_trade =
+            self.calculate_net_curve_trade_from_timestamp(current_block_timestamp)?;
+        // Calculate the LP total supply.
+        let lp_total_supply = active_lp_total_supply + withdrawal_shares_total_supply;
+        let starting_present_value = self.calculate_present_value(current_block_timestamp)?;
+
+        // NOTE: Round the initial guess down to avoid overshooting.
+        //
+        // We make an initial guess for Newton's method by assuming that the
+        // ratio of the share reserves delta to the withdrawal shares
+        // outstanding is equal to the LP share price. In reality, the
+        // withdrawal pool should receive more than this, but it's a good
+        // starting point. The calculation is:
+        //
+        // x_0 = w * (PV(0) / l)
+        let mut share_proceeds =
+            withdrawal_shares_total_supply.mul_div_down(starting_present_value, lp_total_supply);
+
+        // If the pool is net neutral, the initial guess is equal to the final
+        // result.
+        if net_curve_trade == int256!(0) {
+            return Ok(share_proceeds);
+        }
+
+        // Proceed with Newton's method. The objective function, `F(x)`, is
+        // given by:
+        //
+        // F(x) = PV(x) * l - PV(0) * (l - w)
+        //
+        // Newton's method will terminate as soon as the current iteration is
+        // within the minimum tolerance or the maximum number of iterations has
+        // been reached.
+        let mut smallest_delta = I256::zero();
+        let mut closest_share_proceeds = fixed!(0);
+        let mut closest_present_value = fixed!(0);
+        for _ in 0..max(max_iterations, SHARE_PROCEEDS_MAX_ITERATIONS) {
+            // Clamp the share proceeds to the max share reserves delta since
+            // values above this threshold are always invalid.
+            share_proceeds = min(share_proceeds, max_share_reserves_delta);
+
+            // Simulate applying the share proceeds to the reserves.
+            //
+            // NOTE: We are calling this with 'self' so that original values are
+            // used with the updated value of share_proceeds
+            let updated_state =
+                self.get_state_after_liquidity_update(-I256::try_from(share_proceeds)?)?;
```

Comment:
If we want to maintain parity with Solidity, it might be better to `unwrap` instead of bubbling up the `I256::try_from` error. This is just a note since I actually like the way you do it here more, but in Solidity, the entire calculation would be halted.

Reply:
Another way you could do this is to return a different error that indicates to the caller (`calculate_distribute_excess_idle`) that we hit an unrecoverable error. Otherwise, this could make it look like the calculation will succeed when it actually fails.

Reply:
Now that I'm deeper into the review, I like doing this with `?`. I think we just need to be more careful with some of the other functions that return `Result`.

Reply:
Ok I'm resolving this one, sounds like there's nothing to do there.